src/actions/plugins/smartim.ts

```diff
@@ -0,0 +1,93 @@
+import { exec } from 'child_process';
+import { ModeName, Mode } from '../../mode/mode';
+import { existsSync } from 'fs';
+import { configuration } from '../../configuration/configuration';
+import * as vscode from 'vscode';
+
+// SmartIM change input method automatically when mode changed
+export class SmartIM {
+  public savedIM = '';
+
+  public async changeInputMethod(oldMode: ModeName, newMode: ModeName) {
+    const enableSmartIM = configuration.enableSmartIM;
+    if (enableSmartIM !== true) {
+      return;
+    }
+    // when you exit from insert-like mode, save origin input method and set it to default
+    if (
+      oldMode === ModeName.Insert ||
+      oldMode === ModeName.SurroundInputMode ||
+      oldMode === ModeName.Replace
+    ) {
+      if (
+        newMode !== ModeName.Insert &&
+        newMode !== ModeName.SurroundInputMode &&
+        newMode !== ModeName.Replace
+      ) {
+        this.disableOriginInputMethod();
+        return;
+      }
+    }
+    // when you enter insert-like mode, resume origin input method
+    if (
+      newMode === ModeName.Insert ||
+      newMode === ModeName.SurroundInputMode ||
+      newMode === ModeName.Replace
+    ) {
+      if (
+        oldMode !== ModeName.Insert &&
+        oldMode !== ModeName.SurroundInputMode &&
+        oldMode !== ModeName.Replace
+      ) {
+        this.enableOriginInputMethod();
+      }
+    }
+  }
+
+  // save origin input method and set input method to default
+  public async disableOriginInputMethod() {
+    if (process.platform === 'darwin') {
```

Comment:
Does this need to be platform dependent? I know you only tested it on osx, but to I don't see anything that is platform dependent here

Reply:
Woops nevermind, https://github.com/ybian/smartim#why-is-it-mac-only